addon/index.ts

```diff
@@ -192,7 +195,43 @@ const createDecorator = (
  * @param {object?} [options={}]
  * @return {TaskProperty}
  */
-export const task = createDecorator(createTaskFromDescriptor);
+const taskDecorator = createDecorator(createTaskFromDescriptor);
+
+export function task<Args extends any[], R>(
+  taskFn: GeneratorFn<unknown, Args, R>
+): Task<Args, Exclude<R, Promise<any>>>;
+export function task<
+  Args extends any[],
+  R,
+  E extends {
+    // @TODO: this does not work
+    perform: GeneratorFn<E, Args, R>;
+  }
+>(encapsulatedTask: E): Task<Args, Exclude<R, Promise<any>>>;
```

Comment:
Unfortunately this self-referential trickery does not work. ☹️

Reply:
This can also be a blessing in disguise though!

```ts
class Foo {
  @task
  encapsulated = task(class {
    privateState = 123;
    *perform() {
      return this.privateState;
    }
  });
}
```

This would solve https://github.com/machty/ember-concurrency-decorators/issues/37.

Reply:
/cc @spencer516 FYI 😊